File: frontends/android/app/src/test/java/com/example/animastor/repository/ResourceInvalidationsTest.kt
package com.example.animastor.repository

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import kotlinx.coroutines.flow.first
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Generic resource invalidation bus — the Android half of the
 * "external data mutation → invalidation event → cache evict → re-fetch →
 * reactive UI update" pipeline shared conceptually with the web frontend.
 *
 * Pure JVM: verifies dispatch semantics (kinds delivered, resource key
 * contract, monotonic stamp, blank-key rejection) without Android or a device.
 */
class ResourceInvalidationsTest {

    /** Subscribe first, then emit after the subscription had time to attach. */
    private fun awaitNextEvent(): ResourceInvalidations.ResourceEvent = runBlocking {
        val waiter = async(Dispatchers.Default) {
            withTimeout(5000) { ResourceInvalidations.events.first() }
        }
        // Headroom for the subscriber to attach before emitting.
        delay(150)
        ResourceInvalidations.emitExternal(ResourceInvalidations.Keys.book("b1"))
        waiter.await()
    }

    @Test
    fun externalEvent_reachesSubscriber_withScopedResourceKey() {
        val event = awaitNextEvent()
        assertEquals(ResourceInvalidations.Kind.EXTERNAL, event.kind)
        assertEquals("book:b1", event.resource)
    }

    @Test
    fun localEvent_reachesSubscriber_kindIsPreserved() = runBlocking {
        val waiter = async(Dispatchers.Default) {
            withTimeout(5000) { ResourceInvalidations.events.first() }
        }
        delay(150)
        ResourceInvalidations.emitLocal(ResourceInvalidations.Keys.book("b2"))
        val event = waiter.await()
        assertEquals(ResourceInvalidations.Kind.LOCAL, event.kind)
        assertEquals("book:b2", event.resource)
    }

    @Test
    fun stampIsMonotonicAcrossEmissions() {
        val before = ResourceInvalidations.stamp
        ResourceInvalidations.emitExternal(ResourceInvalidations.Keys.book("x"))
        ResourceInvalidations.emitLocal(ResourceInvalidations.Keys.book("x"))
        val after = ResourceInvalidations.stamp
        assertTrue(after > before)
    }

    @Test
    fun blankResourceIsRejected() {
        val before = ResourceInvalidations.stamp
        ResourceInvalidations.emitExternal("")
        assertEquals(before, ResourceInvalidations.stamp)
    }

    @Test
    fun bookKeyFormatIsStable() {
        assertEquals("book:abc123", ResourceInvalidations.Keys.book("abc123"))
        assertTrue(ResourceInvalidations.Keys.book("abc123").startsWith(ResourceInvalidations.Keys.BOOK_PREFIX))
    }

    @Test
    fun localAndExternalAreDistinctKinds() {
        assertNotEquals(ResourceInvalidations.Kind.LOCAL, ResourceInvalidations.Kind.EXTERNAL)
    }
}

File: frontends/app/src/state/resourceInvalidations.test.ts
// Tests for the generic resource invalidation bus (resourceInvalidations.ts) —
// the web half of the "external data mutation → invalidation event → cache
// evict → re-fetch → reactive UI update" pipeline (Android parity:
// ResourceInvalidationsTest.kt). Pure module, no mocks required.
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import {
  bookResource, BOOK_RESOURCE_PREFIX, emitExternal, emitLocal,
  onResourceInvalidated, type ResourceInvalidationEvent,
} from './resourceInvalidations';

describe('resourceInvalidations bus', () => {
  const received: ResourceInvalidationEvent[] = [];
  let unsub: (() => void) | null = null;

  beforeEach(() => {
    received.length = 0;
    unsub = onResourceInvalidated((e) => received.push(e));
  });
  afterEach(() => {
    unsub?.();
    unsub = null;
  });

  it('delivers an EXTERNAL event with the scoped resource key', () => {
    emitExternal(bookResource('b1'));
    expect(received).toEqual([{ kind: 'EXTERNAL', resource: 'book:b1' }]);
  });

  it('delivers a LOCAL event preserving the kind', () => {
    emitLocal(bookResource('b2'));
    expect(received).toEqual([{ kind: 'LOCAL', resource: 'book:b2' }]);
  });

  it('rejects a blank resource key (no event emitted)', () => {
    emitExternal('');
    emitLocal('   ');
    expect(received).toEqual([]);
  });

  it(' unsubscribed listeners no longer receive events', () => {
    unsub?.();
    emitExternal(bookResource('b3'));
    expect(received).toEqual([]);
  });

  it('fan-outs one event to every subscriber', () => {
    const second: ResourceInvalidationEvent[] = [];
    const off = onResourceInvalidated((e) => second.push(e));
    emitExternal(bookResource('b4'));
    expect(received).toEqual(second);
    off();
  });

  it('bookResource key format is stable and prefixed', () => {
    expect(bookResource('abc123')).toBe('book:abc123');
    expect(bookResource('abc123').startsWith(BOOK_RESOURCE_PREFIX)).toBe(true);
  });

  it('does not swallow listener errors of other subscribers (dispatch continues)', () => {
    const err = vi.fn(() => { throw new Error('boom'); });
    const off = onResourceInvalidated(err);
    emitExternal(bookResource('b5'));
    expect(received.length).toBe(1);
    off();
  });
});
